client/circuit_breaker.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"strings"
+
+	"github.com/cloudwego/kitex/client"
+	"github.com/cloudwego/kitex/pkg/circuitbreak"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/kitex-contrib/config-etcd/etcd"
+	"github.com/kitex-contrib/config-etcd/utils"
+)
+
+// WithCircuitBreaker sets the circuit breaker policy from etcd configuration center.
+func WithCircuitBreaker(dest, src string, etcdClient etcd.Client, uniqueID int64, opts utils.Options) []client.Option {
+	param, err := etcdClient.ClientConfigParam(&etcd.ConfigParamConfig{
+		Category:          circuitBreakerConfigName,
+		ServerServiceName: dest,
+		ClientServiceName: src,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	for _, f := range opts.EtcdCustomFunctions {
+		f(&param)
+	}
+	key := param.Prefix + "/" + param.Path
+	cbSuite := initCircuitBreaker(key, dest, src, etcdClient, uniqueID)
+
+	return []client.Option{
+		client.WithCircuitBreaker(cbSuite),
+		client.WithCloseCallbacks(func() error {
+			err := cbSuite.Close()
+			if err != nil {
+				return err
+			}
+			// cancel the configuration listener when client is closed.
+			etcdClient.DeregisterConfig(key, uniqueID)
```

Comment:
Deregisteration should be performed before `cbSuite.Close`

Reply:
Done